File: sync/src/com/kirinpatel/util/FileSelector.java
package com.kirinpatel.util;

import com.kirinpatel.gui.ProgressView;
import com.kirinpatel.net.Server;

import javax.swing.*;
import java.awt.*;
import java.io.*;

public class FileSelector {

    public static File getFile(Component parent) {
        JFileChooser mediaSelector = new JFileChooser("tomcat/webapps/media");
        mediaSelector.setFileSelectionMode(JFileChooser.FILES_ONLY);
        mediaSelector.showOpenDialog(parent);

        File selectedFile = mediaSelector.getSelectedFile();
        if (selectedFile == null) return null;

        if (selectedFile.getAbsolutePath().startsWith(new File("tomcat/webapps/media").getAbsolutePath())) {
            return selectedFile;
        } else {
            return moveFile(selectedFile);
        }
    }

    /**
     * Credit: https://www.mkyong.com/java/how-to-move-file-to-another-directory-in-java/
     *
     * @param selectedFile
     * @return
     */
    private static File moveFile(File selectedFile) {
        File newFile = new File("tomcat/webapps/media/" + selectedFile.getName());

        ProgressView progressView = new ProgressView("Moving media", "Please wait while your media is moved to the proper folder.");
        new Thread(() -> {
            try {
                InputStream inStream = new FileInputStream(selectedFile);
                OutputStream outStream = new FileOutputStream(newFile);

                byte[] buffer = new byte[1024];
                int length;

                while ((length = inStream.read(buffer)) > 0) {
                    outStream.write(buffer, 0, length);
                }

                inStream.close();
                outStream.close();

            } catch(IOException e) {
                e.printStackTrace();
            }
        }).start();

        new Thread(() -> {
            Server.setEnabled(false);
            long timeout = System.currentTimeMillis();
            long time = 0;
            while(newFile.length() < selectedFile.length()) {
                time = System.currentTimeMillis();
                progressView.setProgress(newFile.length(), selectedFile.length());
                if (timeout + (1000 * 300) < time) {
                    new UIMessage("Unable to move media", "Your media was unable to be moved.\nPlease try to manually move your media to the Tomcat directory.", 1);
                    break;
                }
            }

            progressView.dispose();
            Server.setEnabled(true);
            if (timeout + (1000 * 300) > time) new UIMessage("Your media is ready", "Your media has been moved to the\nTomcat folder and is ready for playback.", 0);
        }).start();

        return newFile;
    }
}

File: src/main/java/com/kirinpatel/sync/util/FileSelector.kt
package com.kirinpatel.sync.util

import com.kirinpatel.sync.gui.ProgressView
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import javax.swing.JFileChooser

class FileSelector {
    companion object {
        private var isIterating: Boolean = false
        private val listeners: ArrayList<FileSelectorListener> = ArrayList()

        fun addListener(listener: FileSelectorListener) {
            listeners.add(listener)
        }

        fun removeListener(listener: FileSelectorListener) {
            while (isIterating) {
                Thread.sleep(50)
            }
            listeners.remove(listener)
        }

        fun getFile() : File? {
            val mediaSelector = JFileChooser("tomcat/webapps/media")
            mediaSelector.fileSelectionMode = JFileChooser.FILES_ONLY
            val returnVal: Int = mediaSelector.showOpenDialog(null)
            if (returnVal == JFileChooser.APPROVE_OPTION) {
                val selectedFile = mediaSelector.selectedFile ?: return null
                return if (selectedFile.absolutePath.startsWith(File("tomcat/webapps/media").absolutePath)) {
                    selectedFile
                } else {
                    moveFile(selectedFile)
                }
            } else {
                return null
            }
        }

        private fun moveFile(selectedFile : File) : File {
            val newFile = File("tomcat/webapps/media/" + selectedFile.name)
            var progress: Int = -1
            Thread {
                try {
                    FileInputStream(selectedFile).use { inStream ->
                        FileOutputStream(newFile).use { outStream ->
                            val buffer = ByteArray(1024)
                            var length: Int = inStream.read(buffer)
                            while (length > 0) {
                                outStream.write(buffer, 0, length)
                                length = inStream.read(buffer)
                            }
                        }
                    }
                } catch (e: IOException) {
                    isIterating = true
                    listeners.forEach { listener ->
                        listener.failedToMoveFile(e)
                    }
                    isIterating = false
                }
            }.start()
            ProgressView()
            Thread {
                isIterating = true
                listeners.forEach { listener ->
                    listener.startedMovingFile()
                }
                isIterating = false
                val timeout = System.currentTimeMillis()
                var time: Long
                while (newFile.length() < selectedFile.length()) {
                    time = System.currentTimeMillis()
                    if (progress != (newFile.length() * 100 / selectedFile.length()).toInt()) {
                        isIterating = true
                        listeners.forEach { listener ->
                            listener.fileProgressUpdated((newFile.length() * 100 / selectedFile.length()).toInt())
                        }
                        isIterating = false
                    }
                    if (timeout + 1000 * 300 < time) {
                        isIterating = true
                        listeners.forEach {listener ->
                            listener.failedToMoveFile(IOException("File took too long to move."))
                        }
                        isIterating = false
                        break
                    }
                }
                isIterating = true
                listeners.forEach { listener ->
                    listener.successfullyMovedFile()
                }
                isIterating = false
            }.start()
            return newFile
        }
    }
}